### ISSUE TITLE
Text operations behave incorrectly when remapped

### ISSUE THREAD

**Issue by p-e-w** [martanne/vis#581]
_**Note:** This issue may appear to be superficially similar to #580, but I believe the underlying bug is completely unrelated. Similar problems occur with operations other than "align indent"._

While experimenting with key remapping, I observed various bugs where key presses don't do what they should. Many of these bugs are very complicated to describe and difficult to reproduce. The below represents the minimal example of such a bug I was able to find.

**Steps to reproduce:**

1. Put this in your `visrc.lua`:
```lua
require("vis")
-- Remove all default keybindings
for mode_name, mode in pairs(vis.modes) do
  for key, help in pairs(vis:mappings(mode)) do
    vis:unmap(mode, key)
  end
end
vis:map(vis.modes.NORMAL, "a", "<vis-mode-visual-charwise>")
vis:map(vis.modes.VISUAL, "b", "<vis-cursor-new-lines-below>")
vis:map(vis.modes.VISUAL, "c", "<vis-cursors-align-indent-right>")
```
2. Open a file in Vis that contains
```
abc
def
```
3. Press `a`. Note that you entered visual mode, as desired.
4. Press `b`. Note that a second cursor was created, as desired. There are now cursors on the characters `a` and `d`.
5. Press `c`, ostensibly invoking `CURSORS_ALIGN_INDENT_RIGHT`.

**Expected behavior:** Nothing happens, as the cursors are already right-aligned.

**Actual behavior:** The bottom cursor is indented.

**Comment by martanne** [comment]
First of all I would like to thank you for the high quality reports!

The issue here seem to be that some operations are not really implemented generically enough. Do you have other examples?

Currently `<vis-cursor-new-lines-below>` assumes that it is invoked from normal mode. Hence, it creates cursors instead of selections.

The following patch should fix that particular issue.

```C
diff --git a/main.c b/main.c
index 11ec217..c085e57 100644
--- a/main.c
+++ b/main.c
@@ -1241,6 +1241,7 @@ static const char *repeat(Vis *vis, const char *keys, const Arg *arg) {
 }
 
 static const char *cursors_new(Vis *vis, const char *keys, const Arg *arg) {
+	Text *txt = vis_text(vis);
 	View *view = vis_view(vis);
 	VisCountIterator it = vis_count_iterator_get(vis, 1);
 	while (vis_count_iterator_next(&it)) {
@@ -1274,8 +1275,14 @@ static const char *cursors_new(Vis *vis, const char *keys, const Arg *arg) {
 			else if (arg->i == +1)
 				cursor_new = view_cursors_next(cursor);
 		}
-		if (cursor_new)
+		if (cursor_new) {
+			Filerange sel = view_cursors_selection_get(cursor);
+			if (text_range_valid(&sel)) {
+				sel = text_range_new(newpos, text_char_next(txt, newpos));
+				view_cursors_selection_set(cursor_new, &sel);
+			}
 			view_cursors_primary_set(cursor_new);
+		}
 	}
 	vis_count_set(vis, VIS_COUNT_UNKNOWN);
 	return keys;
```

However, this still has some drawbacks. For example, selections are always created right oriented even when the already existing ones are left extending.

I have local changes which make selections first class primitives. Cursors will be replaced by singleton selections. Normal mode will move both selection endpoints. Whereas in visual mode selections are anchored, only one endpoint will be adjusted, while the other is fixed.

**Comment by p-e-w** [comment]
> Do you have other examples?

Yes, I have observed at least half a dozen oddities trying to get Vis to do what I want, though most of them would require a more elaborate `visrc.lua` to demonstrate.

However, when I think about it I believe that cursors vs. selections are at the heart of all of them. I had noticed for a while that selections seem more like an afterthought in Vis currently (just as they are in Vim), so your idea to merge the two concepts seems really exciting. Hope to see it in the public repository soon!

**Comment by martanne** [comment]
I [pushed my branch](https://github.com/martanne/vis/tree/selections).

A selection is a non-empty, directed range with two endpoints called *cursor* and *anchor*. A selection can be anchored in which case the anchor remains fixed while only the position of the cursor is adjusted. For non-anchored selections both endpoints are updated. A singleton selection covers one character on which both cursor and anchor reside. There always exists a primary selection which remains visible (i.e. changes to its position will adjust the viewport).

Selections can be saved into registers and manipulated with the usual set operations (union, intersect, complement, minus). Additionally they can be pairwise combined (using union, intersection, leftmost, , rightmost, shorter, longer).

This is currently almost completely untested. In particular interaction with visual mode needs to be reviewed. The difference between normal and visual mode should be that in the latter selections are anchored while in the former they are not. Currently this is not really enforced.

Anyway it should hopefully demonstrate where things are heading.

**Comment by martanne** [comment]
Did somebody try the branch? I think the basic stuff should mostly work. Because the changes are quite invasive, I will probably just merge it and then further improve it on master.

Marks and the jump list in particular need to be made aware of multiple selections.

**Comment by erf** [comment]
Seem to work fine. Haven't testet it thoroughly. Maybe you should just merge it and so can people report issues as they show up.

**Comment by martanne** [comment]
This particular issue should be fixed by the selection refactoring and d6c4d8b155de16abcfcffa2a9bbfa886909b3ec3.

A more general issue related to the processing of aliases should be fixed by 54ca598fcccff8844bfbe494cd3b325b2fea487f (albeit in an ugly way).

Please report more issues as you encounter them. Also the cursor related pseudo key names will probably be renamed to use "selection". The Lua API also hasn't yet been updated.
